### PR TITLE
Add support for some JSON-patch operations

### PR DESCRIPTION
**Pull request by d-mart**
* 🔜 Will update docs if the API looks ok
* Modify existing objects using param names and validators as called
  out in a transcoder
* Prevent modification of properties or links by adding a `protected:
  true` option in the transcoder
* "replace" on defproperty and deflink
* "add" (append) to a deflinks list with path: /foo/- syntax
* "replace" entire deflinks list
* JSON-patch RFC: https://tools.ietf.org/html/rfc6902

### CODE REVIEW

**Comment by pezra** [comment]
How would a phoenix action look like for this interface?

File: test/exhal/transcoder_test.exs [hunk 1]
```diff
@@ -238,4 +238,185 @@ defmodule ExHal.TranscoderTest do
...
+    cocktail = MartiniTranscoder.decode!(doc)
+    assert %{spirits: %{liquor: "gin"}} == cocktail
+
+    json_patches = [
+      %{"op" => "replace", "path" => "/baseIngredient", "value" => "vodka"},
+    ]
+
+    bond_style = MartiniTranscoder.patch!(cocktail, json_patches)
+
+    assert %{spirits: %{liquor: "vodka"}} == bond_style
```
Comment (pezra) [inline test/exhal/transcoder_test.exs]: ❌  martinis are made with gin. period. end of story.

😄

**Comment by d-mart** [comment]
> How would a phoenix action look like for this interface?

Something like this (generic/psuedo code, but adapted from a real use)
```elixir
def update(%{method: "PATCH"} = conn, %{"_json" => change_list, "id" => id}) do
  my_thing = Repo.Thing.get(id)  # generic Ecto.Repo lookup
  updated_thing = my_thing
  |> Map.from_struct
  |> ThingTranscoder.patch!(change_list)

  case Repo.Thing.changeset(my_thing, updated_thing) |> Repo.update! do
    {:error, error} -> # use conn to return error
    {:ok, _} -> # use conn to return 204
  end
end
```

File: README.md [hunk 2]
```diff
@@ -231,10 +231,59 @@ iex> PersonTranscoder.encode!(%{name: "Jane Doe",
...
+```
+
+Create an object from the transcoder, then modify it
+
+```elixir
+iex> spot = PetTranscoder.decode!(doc)
+%{name: "Spot",
+  species: "dog",
+  favorite_toy: "http://a.co/56guxwO"
+  friends: ["https://petbook.com/u/fifi", "https://petbook.com/u/fido"]}
```
Comment (pezra) [inline README.md]: Can we change this part of the example to be a proper hal link?  eg:
```
...
  _links: {
    friends: [
      {href: "https://petbook.com/u/fifi"},
      {href: "https://petbook.com/u/fido"}
    ]
  }
}
```

Comment (d-mart) [inline README.md]: In this case the example is a decoded object, not a HAL document.  Like in this example:
```elixir
  test "Object representation test" do
    defmodule AnotherPersonTranscoder do
      use ExHal.Transcoder

      defproperty "name"
      defproperty "mailingAddress", param: :address
      deflink     "app:department", param: :department_url
      deflinks    "tags"
    end

    json = ~s(
      {
        "name": "Jane Doe",
        "mailingAddress": "123 Main St",
        "_links": {
          "app:department": { "href": "http://example.com/dept/42" },
          "tags": [
            {"href": "foo:1"},
            {"href": "http://2"},
            {"href": "urn:1"}
          ]
        }
      }
    )

    json
    |> ExHal.Document.parse!
    |> AnotherPersonTranscoder.decode!
    |> IO.inspect
end
=> %{address: "123 Main St", department_url: "http://example.com/dept/42",
  name: "Jane Doe", tags: ["foo:1", "http://2", "urn:1"]}
```

The pet example could be made into a HAL doc by doing something like this:

```elixir
spot |> PetTranscoder.patch!(json_patches) |> PetTranscoder.encode!
```

Comment (pezra) [inline README.md]: oic... then patch document is wrong, i think.

Given a JSON representation:

```json
{ "name": "Jane Doe",
   "mailingAddress": "123 Main St",
   "_links": {
      "app:department": { "href": "http://example.com/dept/42" },
      "tags": [
        {"href": "foo:1"}
      ]
  }
}
```
If you want to add a tag would be

```json
[ { "op": "add", "path": "/_links/tags/-", "value": { "href": "urn:x" } } ] 
```

Comment (d-mart) [inline README.md]: Good point.  Ok, I will get that in.

**Comment by pezra** [comment]
released as 5.2.1

**Comment by pezra** [comment]
thanks!

**Comment by d-mart** [comment]
Awesome, thanks!

**Pull request merged**